Store tx by Id

ethanfrey opened the pull request:
Depends on #26 
Resolved #24 

* Use txid returned from the api instead of the fakeTxId placeholder
* replace txByHeightAndId with txById
* rename state variable `byHeight` to `byTxId` (which is only a depth one map directly to tx)

This change may break iov-wallet, but not sure. Code in redux should be all properly updated.

rudi-cilibrasi commented on src/prom/accountSample.mjs:
```diff
@@ -89,8 +82,8 @@ async function promAccountTransactionSample(reader, address) {
 
   let transactions;
   try {
+    // TODO: is there any advantage to this? Not just one line result.map(...), no need to copy twice
```
the advantage of the one-per-line format is (like `import`'s) it's easy to add and remove whole lines. i was thinking we will have a small handful of required properties so we can more easily cut and paste just the property we need if we avoid combining lines like you did above.

rudi-cilibrasi commented on src/prom/connectionManager.mjs:
```diff
@@ -2,19 +2,18 @@ import iovCore from '@iov/core';
-    return bnsNetwork;
-  });
+  const networks = chainIds.map(chainId =>
+    bnsConnector(urljoin('wss://bov.'+networkList[chainId].url)));
```
this looks great!

rudi-cilibrasi commented on src/prom/connectionManager.mjs:
```diff
@@ -2,19 +2,18 @@ import iovCore from '@iov/core';
-    writer = new IovWriter(profile, chains);
+    writer = new IovWriter(profile);
+    for (const network of networks) {
+      await writer.addChain(network);
```
is this slower than the old way?  e.g. going in-sequence instead of all-at-once?  @Isabello  may care since i think she wrote the first high-overlap version. i prefer the clarity in this new way but since I know some people love responsive feelings i thought i would mention it

ethanfrey replied:
sure, you are right.
since the async moved to another place, this changed. I can use Promise.all here.

rudi-cilibrasi commented on src/reducer/blockchain/txById.mjs:
```diff
@@ -1,31 +1,30 @@
-        newByHeight[height] = {...newByHeight[height]};
-      }
-      newByHeight[height][txId] = tx;
+      // TODO: should I do the hex-encoding in prom dir? I think there are other Buffers passed around...
```
no because `prom` is more low level and meant for potentially machine use.  the hex-encoding is only necessitated by JSON serializable things like the redux store. we don't need to worry about hex encoding buffers we can't see yet. but we do need to hex encode any buffers used as keys in pojos to avoid destroying the json property.

ethanfrey commented:
Updated with comments, rebased on fixed in iov/core init (pr #26)
Did manual testing and seems to work.

One issue: when I sent 1.2 CASH, my balance dropped from 90 CASH to 89.999999998 CASH. So really 1.000000002 got sent, meaning there is a bug in converting the fractional part into the struct.

ethanfrey commented:
Otherwise, good to merge imo, if you are happy with #26

rudi-cilibrasi commented:
yup the amount bug is longstanding we have been ignoring it awhile merge it! :-)
